builtin/logical/pki/cert_util.go

```diff
@@ -97,24 +133,30 @@ func fetchCertBySerial(req *logical.Request, prefix, serial string) (*logical.St
 // Given a set of requested names for a certificate, verifies that all of them
 // match the various toggles set in the role for controlling issuance.
 // If one does not pass, it is returned in the string argument.
-func validateCommonNames(req *logical.Request, commonNames []string, role *roleEntry) (string, error) {
+func validateNames(req *logical.Request, names []string, role *roleEntry) (string, error) {
 	hostnameRegex, err := regexp.Compile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
 	if err != nil {
-		return "", fmt.Errorf("Error compiling hostname regex: %s", err)
+		return "", fmt.Errorf("error compiling hostname regex: %s", err)
 	}
-	subdomainRegex, err := regexp.Compile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]))*$`)
-	if err != nil {
-		return "", fmt.Errorf("Error compiling subdomain regex: %s", err)
-	}
-	for _, name := range commonNames {
-		if role.AllowLocalhost && name == "localhost" {
-			continue
-		}
-
+	for _, name := range names {
 		sanitizedName := name
+		emailDomain := name
+		isEmail := false
 		isWildcard := false
-		if strings.HasPrefix(name, "*.") {
-			sanitizedName = name[2:]
+		if strings.Contains(name, "@") {
```

Comment:
All the new checks should probably be after `role.AllowAnyName` otherwise they are taking precedence